proxmoxtf/resource/file.go

```diff
@@ -565,14 +572,43 @@ func fileCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag
 			}...)
 		}
 
-		remoteFileDir := *datastore.Path
+		// the temp directory is used to store the file on the node before moving it to the datastore
+		// will be created if it does not exist
+		tempFileDir := fmt.Sprintf("/tmp/tfpve/%s", uuid.NewString())
 
-		err = capi.SSH().NodeUpload(ctx, nodeName, remoteFileDir, request)
-	}
+		err = capi.SSH().NodeUpload(ctx, nodeName, tempFileDir, request)
+		if err != nil {
+			diags = append(diags, diag.FromErr(err)...)
+			return diags
+		}
 
-	if err != nil {
-		diags = append(diags, diag.FromErr(err)...)
-		return diags
+		// handle the case where the file is uploaded to a subdirectory of the datastore
+		srcDir := tempFileDir
+		dstDir := *datastore.Path
+
+		if request.ContentType != "" {
+			srcDir = tempFileDir + "/" + request.ContentType
+			dstDir = *datastore.Path + "/" + request.ContentType
+		}
+
+		_, err := capi.SSH().ExecuteNodeCommands(ctx, nodeName, []string{
```

Comment:
Glang's ssh has [Session.Stdin](https://pkg.go.dev/golang.org/x/crypto/ssh#Session), it might be good idea to expose that option using something like `ExecuteNodeCommandsWithInput`. Then something like this:
```
ExecuteNodeCommandsWithInput(ctx, nodeName, uploadedFileContents, []string{
        fmt.Sprintf(`%s; try_sudo "/usr/bin/tee %s/%s"`, trySudo, dstDir, *fileName),
});
```
could place file directly to the target datastore. (Probably also connecting `Stdout` to `nil`, to avoid the output of the `tee` command)

The upload to temp directory does create directories as [needed](https://github.com/bpg/terraform-provider-proxmox/blob/78fa7b1a4f4399948f865522f2b1ce80b062daf4/proxmox/ssh/client.go#L198), but it does not modify default permissions, it looks like everything would be *world-readable by default* (even though the file would be there only for a very short moment).

Non-root user with ssh access could use something like `inotify`, to watch `/tmp` for new files and copy their contents away, getting hold of other users' cloud-init snippet data, which can contain secrets like passwords/keys to access other services.

Ideally, each ssh user's access would be limited to snippets the datastores based on `sudoers` file and/od datastore directory permissions. The use of temp directory would allow users to eavesdrop on all snippets in flight.

Reply:
Thanks, good points! I haven't thought that deep 😅
I'll put it as a future work in a separate ticket.